### PR TITLE
JSAPIObject and Internal Fields

### PR DESCRIPTION
**Pull request by indutny**
see #50 

cc @hhellyer @yjhjstz

### CODE REVIEW

**Comment by indutny** [comment]
I'll write a test in a bit.

**Comment by indutny** [comment]
Pushed a test.

**Comment by indutny** [comment]
@nodejs/jenkins-admins we really need to have a CI coverage for this repo. Please help us!

File: src/llv8-constants.cc [hunk 1]
```diff
@@ -472,6 +500,8 @@ void Types::Load() {
       LoadConstant("type_JSGlobalObject__JS_GLOBAL_OBJECT_TYPE");
   kOddballType = LoadConstant("type_Oddball__ODDBALL_TYPE");
   kJSObjectType = LoadConstant("type_JSObject__JS_OBJECT_TYPE");
+  kJSAPIObjectType = LoadConstant("APIObjectType");
```
Comment (indutny) [inline src/llv8-constants.cc]: I'm going to send a CL to v8 team for these constants.

**Comment by indutny** [comment]
V8 CL: https://codereview.chromium.org/2514063002/

File: src/llv8.cc [hunk 2]
```diff
@@ -1179,13 +1179,52 @@ std::string JSObject::Inspect(InspectOptions* options, Error& err) {
...
+  int64_t instance_size = map.InstanceSize(err);
+
+  // kVariableSizeSentinel == 0
+  // TODO(indutny): post-mortem constant for this?
+  if (err.Fail() || instance_size == 0) return std::string();
+
+  std::string res;
+  for (int64_t off = v8()->js_object()->kInternalFieldsOffset;
+       off < instance_size; off += v8()->common()->kPointerSize) {
+    int64_t field = LoadField(off, err);
```
Comment (yjhjstz) [inline src/llv8.cc]: seem we can't inspect internal fieldd in a recursive way.

Comment (indutny) [inline src/llv8.cc]: Internal field is a pointer to C++ object. We can't really do it.

File: src/llv8-constants.cc (hunk 1 above)
Comment (indutny) [inline src/llv8-constants.cc]: Sent.

**Comment by indutny** [comment]
Does it LGTY @yjhjstz (and @hhellyer )?

**Comment by rvagg** [comment]
@indutny whatchu need that travis isn't providing? GIve us a baseline of platform requirements and we'll see what we can hook up.

**Comment by indutny** [comment]
@rvagg I need just travis, actually, but we can't enable it without giving access to all private repos in this org, AFAIK.

**Comment by indutny** [comment]
@yjhjstz @hhellyer pushed tests for llscan here too. PTAL.

**Comment by indutny** [comment]
CI: https://ci.nodejs.org/job/llnode/9/

**Comment by indutny** [comment]
Internal field inspection doesn't work that well atm, as it account in-object properties too. Will look into fixing it.

**Comment by indutny** [comment]
Added fix to account for in-object properties.

**Comment by hhellyer** [comment]
Changes look good but it is worth adding the code that uses the extra constants from gen-postmortem-metadata.py before that patch has made it into v8 (and then into node)?
Would it be better to split those out from the scan tests and put the tests in under a separate PR?

**Comment by indutny** [comment]
@hhellyer that V8 CL is about to land (just got LGTM), and I think it is quite usual practice in this repo to introduce them at the same time if fallback is possible.

**Comment by indutny** [comment]
@yjhjstz does it LGTY?

**Comment by indutny** [comment]
@hhellyer CL landed, I don't really feel like splitting PR at this point. Is it ok?

**Comment by indutny** [comment]
Landed in ad45a96, 4459f90, and b7b0306. Thank you!

**Comment by indutny** [comment]
We can always revert these changes if anything will go wrong.

**Pull request closed without merging**